src/models/document/base-document-content.ts

```diff
@@ -1002,215 +1004,153 @@ export const BaseDocumentContentModel = types
       });
 
       // Add copied tiles to document
-      const results = self.copyTilesIntoNewRows(updatedTiles, rowIndex);
+      const results = insertTileFunction(updatedTiles, rowInfo);
 
       // Increment default titles when necessary
       results.forEach((result, i) => {
-        const newTile = result?.tileId && self.getTile(result.tileId);
-        if (result && newTile) {
-          const tileContentInfo = getTileContentInfo(newTile.content.type);
-          if (tileContentInfo) {
-            const match = titleMatchesDefault(newTile.title, tileContentInfo.titleBase);
-            if (match) {
-              const newTitle = self.getNewTileTitle(newTile.content.type);
+        if (result?.tileId) {
+          const { oldTitle, newTitle } = self.updateDefaultTileTitle(result.tileId);
 
-              // If the tile title needed to be update, we assume we should also update the data set's name
-              newSharedModelEntries.forEach(sharedModelEntry => {
-                const oldName = sharedModelEntry.sharedModel.dataSet.name;
-                if (oldName === newTile.title) {
-                  sharedModelEntry.sharedModel.dataSet.name = newTitle;
+          // If the tile title needed to be updated, we assume we should also update the data set's name
+          if (newTitle && sharedModelEntries) {
+            newSharedModelEntries.forEach(sharedModelEntry => {
+              if (isSharedDataSetSnapshot(sharedModelEntry.sharedModel)) {
+                const sharedDataSet = sharedModelEntry.sharedModel as SharedDataSetSnapshotType;
+                const oldName = sharedDataSet.dataSet.name;
+                if (oldName === oldTitle) {
+                  sharedDataSet.dataSet.name = newTitle;
                 }
-              });
-
-              newTile.setTitle(newTitle);
-            }
+              }
+            });
           }
         }
       });
 
       // Update tile ids for shared models and add copies to document
       newSharedModelEntries.forEach(sharedModelEntry => {
-        const updatedTileIds: string[] = sharedModelEntry.tiles.map((oldTile: any) => tileIdMap[oldTile.id]);
-        const updatedSharedModel = { ...sharedModelEntry.sharedModel };
-        if (sharedModelEntry.sharedModel.type === "SharedDataSet") {
-          updatedSharedModel.providerId = tileIdMap[sharedModelEntry.sharedModel.providerId];
+        const updatedTileIds: string[] = sharedModelEntry.tiles.map((oldTile: PartialTile) => tileIdMap[oldTile.id])
+          .filter((tileId: string | undefined) => tileId !== undefined);
+        if (isSharedDataSetSnapshot(sharedModelEntry.sharedModel)) {
+          const updatedSharedModel = { ...sharedModelEntry.sharedModel };
+          updateSharedDataSetSnapshotWithNewTileIds(updatedSharedModel as SharedDataSetSnapshotType, tileIdMap);
+          const newSharedModelEntry =
+            self.addSharedModel(SharedDataSet.create(updatedSharedModel as SharedDataSetSnapshotType));
```

Comment:
You can eliminate a couple more casts with:
```suggestion
          const updatedSharedModel: SharedDataSetSnapshotType = { ...sharedModelEntry.sharedModel };
          updateSharedDataSetSnapshotWithNewTileIds(updatedSharedModel, tileIdMap);
          const newSharedModelEntry =
            self.addSharedModel(SharedDataSet.create(updatedSharedModel));
```